File: 402. Remove K Digits.cpp
class Solution {
public:
    string removeKdigits(string num, int k) {
        if(k == 0) return num;

        if(k >= num.length()) return "0";

        stack<char> st;

        for(char x : num) {
            while(!st.empty() && k > 0 && st.top() > x)
                st.pop(), k--;
            
            st.push(x);

            if(st.size() == 1 && st.top() == '0') st.pop();
        }

        while(k-- && !st.empty()) st.pop();

        string ans;
        while(!st.empty())
            ans.push_back(st.top()), st.pop();
        
        reverse(ans.begin(), ans.end());

        return ans.empty() ? "0" : ans;
    }
};

File: 402. Remove K Digits.py
class Solution:
    def removeKdigits(self, num: str, k: int) -> str:
        if(k == 0): 
            return num
        
        if(k >= len(num)):
            return "0"

        st = []

        for x in num:
            while(len(st) > 0 and k > 0 and st[-1] > x):
                st.pop()
                k -= 1
            
            st.append(x)

            if(len(st) == 1 and st[-1] == '0'):
                st.pop()
        
        while(k > 0 and len(st) > 0):
            st.pop()
            k -= 1
        
        ans = ''.join(st)

        return "0" if len(ans) == 0 else ans
